File: 8.回溯/216.组合总和-iii.py
#
# @lc app=leetcode.cn id=216 lang=python3
#
# [216] 组合总和 III
#
from mytools import *
# @lc code=start
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res = []
        def dfs(n, k, sta, pack: List, sums):
            if len(pack) == k:
                if sums == n:
                    nonlocal res
                    res.append(pack.copy())
                return
            for i in range (sta, 9 - (k - len(pack)) + 2):
                sums += i
                if sums > n:
                    sums -= i
                    break
                pack.append(i)
                dfs(n, k, i + 1, pack, sums)
                sums -= i
                pack.pop()
            return
        dfs(n, k, 1, [], 0)
        return res

# @lc code=end

File: 8.回溯/216.组合总和-iii.cpp
/*
 * @lc app=leetcode.cn id=216 lang=cpp
 *
 * [216] 组合总和 III
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        vector<int> pack;
        dfs(res, pack, 1, k, 0, n);
        return res;
    }
    void dfs(vector<vector<int>>& res, vector<int>& pack, int sta, int k, int sums, int n)
    {
        if (pack.size() == k)
        {
            if (sums == n)
            {
                res.emplace_back(pack);
            }
            return;
        }
        for (int i = sta; i <= 9; i ++ )
        {
            sums += i;
            if (sums > n)
            {
                sums -= i;
                break;
            }
            pack.push_back(i);
            dfs(res, pack, i + 1, k, sums, n);
            pack.pop_back();
            sums -= i;
        }
    }
};
// @lc code=end

